spec/extensions/linear_algebra_functions.md

```diff
@@ -465,13 +465,13 @@ Computes the (Moore-Penrose) pseudo-inverse of a matrix (or a stack of square ma
 (function-linalg-qr)=
 ### linalg.qr(x, /, *, mode='reduced')
 
-Computes the qr factorization of a matrix (or a stack of matrices), where `q` is an orthonormal matrix (or a stack of matrices) and `r` is an upper-triangular matrix (or a stack of matrices).
+Computes the qr factorization of a full column rank matrix (or a stack of matrices), where `q` is an orthonormal matrix (or a stack of matrices) and `r` is an upper-triangular matrix (or a stack of matrices).
```

Comment:
[Same problem as in removing "upper"](https://github.com/data-apis/array-api/issues/217#issuecomment-888029418): This "full column rank" addition seems to suggest an array library would do an expansive check before sending the input to the solver. I think in today's call we wanted to rephrase it for the other case such that this interpretation can be avoided (@kgryte will probably update us), so maybe we need to consider a similar phrasing here as well.

Reply:
Note that that is a slippery slope. If you take that route you may also want to reword [linalg.solve](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html?highlight=linalg%20solve#linalg-solve-x1-x2) which assumes that the input is full-rank. Same for [linalg.cholesky](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html?highlight=linalg%20solve#linalg-cholesky-x-upper-false) which takes a symmetric positive definite matrix (SPD).

In my opinion, it is perfectly fine to left the array API unspecified for some inputs. That does not mean that the libraries will have to check that the input is well-formed explicitly. This becomes obvious when one tries to implements some of the checks. For example, a way to check whether a matrix is SPD is to perform Cholesky and see if it succeeds. In the same fashion, a way to see if a matrix is full rank is to perform a QR decomposition and see if it succeeds.

Furthermore, leaving unspecified does not mean that the libraries will have to check that the inputs are correct upfront. For example, in PyTorch we are looking into allowing rectangular inputs for `solve`, and these being dispatched directly into `lstsq`. If the rectangular input case is left unspecified, this is a conformant implementation of the API. On the other hand, if the API specifies that the input should have this shape (or these properties) **or throw an error** that would be much more problematic.

Finally, note leaving things unspecified is a common practice in other standards, such as that of C++, where not even the size of ints is specified.

Reply:
A path forward which can assuage both concerns is

1.  Accept the proposed wording change. This would be in-line with `linalg.cholesky` et al in which certain assumptions are stated explicitly.
2.  Add a note to the specification stating that whether an array library explicitly checks whether an input array is full column rank is implementation-defined. Meaning, an array library can choose whether to check or not before passing to solver.
3.  Add a note to the specification stating that how an array library handles input arrays which do not fulfill assumptions (e.g., full column rank) is implementation-defined.

Reply:
Will address stated assumptions in a follow-up PR in accordance with my comment above.

Reply:
Addressed in https://github.com/data-apis/array-api/pull/302.